src/utils/vaultMath.ts

```diff
@@ -92,36 +92,118 @@ export const debtAfterDelta = (
 
     return AmountMath.add(
       totalDebt,
-      ceilMultiplyBy(
-        AmountMath.make(totalDebt.brand, debtDeltaValue),
-        loanFeeMultiplier,
-      ),
+      ceilMultiplyBy(debtChange, loanFeeMultiplier),
     );
   }
 
-  if (debtDeltaValue <= totalDebt.value) {
-    return AmountMath.make(totalDebt.brand, totalDebt.value - debtDeltaValue);
+  if (AmountMath.isGTE(totalDebt, debtChange)) {
+    return AmountMath.subtract(totalDebt, debtChange);
   }
 
   return AmountMath.makeEmpty(totalDebt.brand);
 };
 
-export const lockedAfterDelta = (
+export const lockedAfterChange = (
   collateralAction: CollateralAction,
   locked: Amount<'nat'>,
-  collateralDeltaValue: bigint | null,
+  lockedChange: Amount<'nat'> | null,
 ): Amount<'nat'> => {
-  if (collateralAction === CollateralAction.None || !collateralDeltaValue) {
+  if (collateralAction === CollateralAction.None || !lockedChange?.value) {
     return locked;
   }
 
   if (collateralAction === CollateralAction.Deposit) {
-    return AmountMath.make(locked.brand, locked.value + collateralDeltaValue);
+    return AmountMath.add(locked, lockedChange);
   }
 
-  if (collateralDeltaValue <= locked.value) {
-    return AmountMath.make(locked.brand, locked.value - collateralDeltaValue);
+  if (AmountMath.isGTE(locked, lockedChange)) {
+    return AmountMath.subtract(locked, lockedChange);
   }
 
   return AmountMath.makeEmpty(locked.brand);
 };
+
+export const istAvailable = (
+  debtLimit: Amount<'nat'>,
+  totalDebt: Amount<'nat'>,
+): Amount<'nat'> =>
+  AmountMath.isGTE(debtLimit, totalDebt)
+    ? AmountMath.subtract(debtLimit, totalDebt)
+    : AmountMath.makeEmpty(debtLimit.brand);
+
+export const maxCollateralForNewVault = (
+  debtLimit: Amount<'nat'>,
+  totalDebt: Amount<'nat'>,
+  loanFee: Ratio,
+  price: PriceDescription,
+  desiredCollateralization: Ratio,
+  collateralPurseBalance: Amount<'nat'>,
+): bigint => {
+  const istAvailableAfterLoanFee = istAvailable(debtLimit, totalDebt);
+
+  const loanFeeMultiplier = addRatios(
+    loanFee,
+    makeRatioFromAmounts(loanFee.denominator, loanFee.denominator),
+  );
+
+  const istAvailableBeforeLoanFee = floorDivideBy(
+    istAvailableAfterLoanFee,
+    loanFeeMultiplier,
+  );
+
+  const collateralForAvailableIst = computeToLock(
+    makeRatioFromAmounts(price.amountOut, price.amountIn),
+    desiredCollateralization,
+    istAvailableBeforeLoanFee.value,
+    desiredCollateralization,
+    loanFee,
+  );
+
+  return AmountMath.min(
+    AmountMath.make(collateralPurseBalance.brand, collateralForAvailableIst),
+    collateralPurseBalance,
+  ).value;
```

Comment:
consider returning the Amount, so at the callsite it's clear that it cannot be neg.

ditto for other amount returns